joint_limits_controller/src/joint_limits_controller.cpp

```diff
@@ -0,0 +1,124 @@
+#include <joint_limits_controller/limited_joint_handle.h>
+#include <urdf/model.h>
+#include <ros/ros.h>
+
+#include <controller_interface/controller_base.h>
+#include <pluginlib/class_loader.h>
+
+namespace joint_limits_controller {
+
+class JointLimitsController : public controller_interface::ControllerBase {
+    pluginlib::ClassLoader<controller_interface::ControllerBase> loader_;
+    boost::shared_ptr<controller_interface::ControllerBase> controller_;
+    std::vector<boost::shared_ptr<LimitedJointHandle> > limit_handles_;
+
+    template<typename T> bool track(const hardware_interface::JointHandle &handle, ros::NodeHandle& root_nh, ros::NodeHandle &controller_nh) {
+        std::string name = handle.getName();
+        LimitedJointHandle::Limits limits(urdf_.getJoint(name));
+        limits.merge(name, root_nh, true);
+        limits.merge(name, controller_nh, true);
+        boost::shared_ptr<LimitedJointHandle> ptr(new T(handle, limits));
+        limit_handles_.push_back(ptr);
+        return true; // TODO
+    }
+
+    template<typename T> bool track(hardware_interface::JointCommandInterface *interface, ros::NodeHandle& root_nh, ros::NodeHandle &controller_nh, std::set<std::string>& claimed_resources){
+        if(!interface) {
+            ROS_ERROR("Interface invalid");
+            return false;
+        }
+        for(std::set<std::string>::iterator it = claimed_resources.begin(); it != claimed_resources.end(); ++it){
+            if(!track<T>(interface->getHandle(*it), root_nh, controller_nh)) return false;
+        }
+        return true;
+    }
+
+    urdf::Model urdf_;
+public:
+    JointLimitsController()
+    : loader_("controller_interface", "controller_interface::ControllerBase")
+    {}
+
+    virtual std::string getHardwareInterfaceType() const { return controller_ ? controller_->getHardwareInterfaceType() : std::string(); }
+
+    virtual bool initRequest(hardware_interface::RobotHW* hw, ros::NodeHandle& root_nh, ros::NodeHandle &controller_nh, std::set<std::string>& claimed_resources){
+
+        if (state_ != CONSTRUCTED){
+            ROS_ERROR("Controller is not in CONSTRUCTED state.");
+            return false;
+        }
+
+        std::string controller_type;
+        if(!controller_nh.getParam("type", controller_type)){
+            ROS_ERROR("Missing controller param");
+            return false;
+        }
+        if(controller_type == "joint_limits_controller::JointLimitsController"){
+            if(!controller_nh.getParam("controller", controller_type)){
+                ROS_ERROR("Missing controller param");
+                return false;
+            }
+        }
+
+        if(!loader_.isClassAvailable(controller_type)){
+            ROS_ERROR_STREAM("Controller type " << controller_type << " is not available");
+            return false;
+        }
+
+        std::string has_robot_description;
```

Comment:
This is string for the absolute name of the `has_robot_description` parameter, I will find another name fot it.